### DISCUSSION TITLE
How can I get text copied to clipboard in headless mode

### DISCUSSION THREAD

**Discussion by ReshailLean** [webdriverio/webdriverio#10810] [Q&A] [answered]
I am struggling to get the copied text from the clipboard in Webdriverio (headless). I got it to work in Puppeteer but a similar approach does not work in Webdriverio

Do you guys have any suggestions on how to make it work headless? 

**puppeteer example:**
const browser = await puppeteer.launch();
const context = browser.defaultBrowserContext();
await context.overridePermissions(/* browser origin */, ['clipboard-read']);
const page = await browser.newPage();
await page.goto('https://mypage.com');
await page.click('#my-component');
console.log(await page.evaluate(() => navigator.clipboard.readText()))

**What I've tried in webdriverio:**
const clipboardText = await browser.execute(() => navigator.clipboard.read())
console.log("copied text: " + clipboardText)

But I am getting this error: javascript error: Read permission denied.

**Comment by ReshailLean** [comment]
I tried the solution from this [thread](https://github.com/webdriverio/webdriverio/discussions/9361)

But it returns a permission error. 

`{"ABORT_ERR":20,"DATA_CLONE_ERR":25,"DOMSTRING_SIZE_ERR":2,"HIERARCHY_REQUEST_ERR":3,"INDEX_SIZE_ERR":1,"INUSE_ATTRIBUTE_ERR":10,"INVALID_ACCESS_ERR":15,"INVALID_CHARACTER_ERR":5,"INVALID_MODIFICATION_ERR":13,"INVALID_NODE_TYPE_ERR":24,"INVALID_STATE_ERR":11,"NAMESPACE_ERR":14,"NETWORK_ERR":19,"NOT_FOUND_ERR":8,"NOT_SUPPORTED_ERR":9,"NO_DATA_ALLOWED_ERR":6,"NO_MODIFICATION_ALLOWED_ERR":7,"QUOTA_EXCEEDED_ERR":22,"SECURITY_ERR":18,"SYNTAX_ERR":12,"TIMEOUT_ERR":23,"TYPE_MISMATCH_ERR":17,"URL_MISMATCH_ERR":21,"VALIDATION_ERR":16,"WRONG_DOCUMENT_ERR":4,"code":0,"message":"Read permission denied.","name":"NotAllowedError"}`

**Comment by ReshailLean** [comment, marked as answer]
I've found a solution not sure if there is a better way to do this but you can get the puppeteer browser instance and use similar code to what I've attached in the original question to get the text from clipboard. 

`await browser.getPuppeteer()` -> https://webdriver.io/docs/api/browser/getPuppeteer/

**Comment by AlanShi01** [comment]
Hello @ReshailLean ,

I tried the following code, but clipboardText is "". Can you please help me with it? Thank you.
 
    await browser.url('http://localhost:8080');
    const puppeteerBrowser = await browser.getPuppeteer();
    // switch to Puppeteer
    const clipboardText = await browser.call(async () => {
        const context = puppeteerBrowser.browserContexts()[0];
        await context.overridePermissions(await browser.getUrl(), ['clipboard-read']);
        const pages = await puppeteerBrowser.pages();
        
        pages[0].click('button[aria-label="copy-to-clipboard"]');

        const clipboardText = (await pages[0].evaluate(() => navigator.clipboard.readText()));
        console.log("copied text: " + clipboardText);

        return clipboardText;
    })

    console.log(clipboardText) // returns ""

config is as follows

      capabilities: [
      {
     browserName: 'chrome',
     'goog:chromeOptions': {
      args:[
         '--headless=new',
         'disable-infobars',
         'disable-popup-blocking',
         'disable-notifications'
     ],
     prefs: {
         'profile.managed_default_content_settings.popups' : 2,
         'profile.managed_default_content_settings.notifications' : 2,
     }
     }
     }

**Comment by ReshailLean** [comment]
Hi @AlanShi01 
This is what works for me. I am using wdio V7 (not that it should matter)

`  async getTextFromClipboard () {
    console.log('=== GET TEXT FROM CLIPBOARD ===')
    const puppeteerBrowser = await browser.getPuppeteer()
    const pages = await puppeteerBrowser.pages()
    const context = puppeteerBrowser.defaultBrowserContext()
    await context.overridePermissions(yourURL, ['clipboard-read'])
    return await pages[0].evaluate(() => navigator.clipboard.readText())
  }`
